torch/csrc/api/include/torch/nn/parallel/data_parallel.h

```diff
@@ -24,6 +25,115 @@
 
 namespace torch {
 namespace nn {
+
+namespace {
+
+// Note [Replicating Modules]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
+//
+// Module replication is implemented in the following two steps:
+// 1) create a module replica on each destination device using Module.clone().
+// 2) manually add a gradient edge pointing from every parameter X in every
+//    module replica to the same parameter X in the original module, using
+//    ReduceAdd as the grad_fn.
+//
+// ReduceAdd can ONLY be used during the backward pass of data parallel. Forward
+// pass cannot use this function as it does not setup gradient function and
+// history at all. Do NOT try to use ReduceAdd for any other purposes.
```

Comment:
Do we support double backwards through replicate in Python? I don't see any indication in issues that we don't...

Reply:
Haven't tried yet. Let me add a test.